### ISSUE TITLE
Configuration cache fails to load when placing a ValueSource in a MapProperty.

### ISSUE THREAD

**Issue by cloudshiftchris** [gradle/gradle#22331]
### Expected Behavior

Configuration cache loads successfully when ValueSource items are contained in MapProperty

### Current Behavior

Configuration Cache fails to load with this exception:

```

org.gradle.api.GradleException: Could not load the value of field `transformer` of `org.gradle.api.internal.provider.TransformBackedProvider` bean found
    in field `providers` of `org.gradle.api.internal.provider.DefaultMapProperty$CollectingProvider` bean
    found in field `__mapProp__` of task `:myTask` of type `Build_gradle$MyTask`.
	at org.gradle.configurationcache.serialization.beans.BeanPropertyReaderKt.readPropertyValue(BeanPropertyReader.kt:108)
	at org.gradle.configurationcache.serialization.beans.BeanPropertyReader.readStateOf(BeanPropertyReader.kt:67)
	at org.gradle.configurationcache.serialization.codecs.BeanCodec.readBeanOf(BeanCodec.kt:72)
	at org.gradle.configurationcache.serialization.codecs.BeanCodec.decode(BeanCodec.kt:47)
	at org.gradle.configurationcache.serialization.codecs.BindingsBackedCodec.decode(BindingsBackedCodec.kt:59)
	at org.gradle.configurationcache.serialization.codecs.FixedValueReplacingProviderCodec.decodeValue(ProviderCodecs.kt:119)
	at org.gradle.configurationcache.serialization.codecs.FixedValueReplacingProviderCodec.decodeProvider(ProviderCodecs.kt:108)
	at org.gradle.configurationcache.serialization.codecs.ProviderCodec.decode(ProviderCodecs.kt:139)
	at org.gradle.configurationcache.serialization.codecs.BindingsBackedCodec.decode(BindingsBackedCodec.kt:59)
	at org.gradle.configurationcache.serialization.DefaultReadContext.read(Contexts.kt:259)
	at org.gradle.configurationcache.serialization.CombinatorsKt.readCollectionInto(Combinators.kt:238)
	at org.gradle.configurationcache.serialization.codecs.CollectionCodecsKt$collectionCodec$2.invokeSuspend(CollectionCodecs.kt:137)
	at org.gradle.configurationcache.serialization.codecs.CollectionCodecsKt$collectionCodec$2.invoke(CollectionCodecs.kt)
	at org.gradle.configurationcache.serialization.codecs.CollectionCodecsKt$collectionCodec$2.invoke(CollectionCodecs.kt)
	at org.gradle.configurationcache.serialization.CombinatorsKt$codec$1.decode(Combinators.kt:90)
	at org.gradle.configurationcache.serialization.codecs.BindingsBackedCodec.decode(BindingsBackedCodec.kt:59)
	at org.gradle.configurationcache.serialization.DefaultReadContext.read(Contexts.kt:259)
	at org.gradle.configurationcache.serialization.beans.BeanPropertyReaderKt.readPropertyValue(BeanPropertyReader.kt:102)
	at org.gradle.configurationcache.serialization.beans.BeanPropertyReader.readStateOf(BeanPropertyReader.kt:67)
	at org.gradle.configurationcache.serialization.codecs.BeanCodec.readBeanOf(BeanCodec.kt:72)
	at org.gradle.configurationcache.serialization.codecs.BeanCodec.decode(BeanCodec.kt:47)
	at org.gradle.configurationcache.serialization.codecs.BindingsBackedCodec.decode(BindingsBackedCodec.kt:59)
	at org.gradle.configurationcache.serialization.codecs.FixedValueReplacingProviderCodec.decodeValue(ProviderCodecs.kt:119)
	at org.gradle.configurationcache.serialization.codecs.MapPropertyCodec.decode(ProviderCodecs.kt:351)
	at org.gradle.configurationcache.serialization.codecs.BindingsBackedCodec.decode(BindingsBackedCodec.kt:59)
	...
Caused by: java.lang.ClassNotFoundException: org.gradle.api.internal.provider.MapCollectors$EntryWithValueFromProvider$$Lambda$2491/0x0000000801b1ea28
	at org.gradle.configurationcache.serialization.DefaultReadContext.readClass(Contexts.kt:285)
	at org.gradle.configurationcache.serialization.codecs.BeanCodec.decode(BeanCodec.kt:44)
	at org.gradle.configurationcache.serialization.CombinatorsKt$reentrant$1$decodeLoop$1.invokeSuspend(Combinators.kt:165)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlin.coroutines.ContinuationKt.startCoroutine(Continuation.kt:115)
	at org.gradle.configurationcache.serialization.CombinatorsKt$reentrant$1.decodeLoop(Combinators.kt:166)
	at org.gradle.configurationcache.serialization.CombinatorsKt$reentrant$1.decode(Combinators.kt:130)
	at org.gradle.configurationcache.serialization.codecs.BindingsBackedCodec.decode(BindingsBackedCodec.kt:59)
	at org.gradle.configurationcache.serialization.DefaultReadContext.read(Contexts.kt:259)
	at org.gradle.configurationcache.serialization.beans.BeanPropertyReaderKt.readPropertyValue(BeanPropertyReader.kt:102)
```

### Context

Unable to store common ValueSource implementations such as environment variables, system properties or Gradle properties in a MapProperty.
For example, this will cause a failure:
`mapProp.put("foo", providers.systemProperty("user.name"))`

However, this works:

`mapProp.put("foo", provider { System.getProperty("user.name") })`

### Steps to Reproduce 

Reproduction project available [here](https://github.com/cloudshiftchris/gradle-configuration-cache-map-property-repro).

### Your Environment

Gradle 7.5.1, MacOS, Java 17

**Comment by cloudshiftchris** [comment]
Note there are workarounds for environment variables and system properties; no such luck for other ValueSource elements, such as Gradle properties.

**Comment by lacasseio** [comment]
This is also the case for `ListProperty` and `SetProperty`. It seems the [`ListProperty` and `SetProperty` were fixed in Gradle 7.5](https://github.com/gradle/gradle/commit/2a81b209c500ddb46128a5d5ece326710f8f318e) and [`MapProperty` will be fixed in Gradle 7.6/8.0](https://github.com/gradle/gradle/commit/6586c031b9ca5e76aa01a3519f6ef5fed55a70d5).

Here are the considerations for the users that need to support older Gradle and hit this issue. The issue arises because adding single elements to a list/set/map is normalized into a single-element iterable or map via mapping. Unfortunately, the normalization maps the element provider using a lambda which was not marked as `Serializable`. During the configuration cache serialization, Gradle sees a `TransformBackedProvider` and happily serializes the lambda but won't be able to deserialize the lambda later. To solve this issue on older Gradle, avoid adding single-element providers to list/set/map property. Instead, always use `addAll` or `putAll`. For those single-element providers, you will need to normalize them manually; at this point, make sure to use a serializable transformer. :-) Groovy closures are already serializable, I'm unsure about Kotlin. In Java you can force a transformer to be serializable via casting: `(Transformer<OutputType, InputType> & Serializable) it -> ...`. You can also use [Gradle's trick](https://github.com/gradle/gradle/blob/38d23184b0e5a7a4d35b8670fd4cec63c4769921/subprojects/base-services/src/main/java/org/gradle/api/internal/lambdas/SerializableLambdas.java#L45), make sure you don't use that utility class given its internal API (make a copy in your code if licensing allows it).

**Comment by lacasseio** [comment]
I would go as far as suggesting using a utility method when supporting older Gradle, something along the line of:
```
<T> void add(HasMultipleValues<T> self, Provider<? extends T> provider) {
    if (isGradleAffectedByIssue2231()) {
        self.addAll(provider.map((Transformer<Iterable<T>, T> & Serializable) it -> Collections.singletonList(it)))
    } else {
        self.add(provider)
    }
}
```
This makes the code base easier to maintain. When your minimum support becomes 7.5+/7.6+, you can refactor all usage of the utility method to the plain Gradle API. The method `isGradleAffectedByIssue2231()` should use `GradleVersion#current()` and `GradleVersion.version(String)` to detect affected Gradle versions.

**Comment by mlopatkin** [comment]
This is fixed in Gradle 8.0.
